controllers/s3utils.go

```diff
@@ -164,19 +192,49 @@ func getS3Secret(ctx context.Context, r client.Reader,
 	return
 }
 
+type s3ObjectStore struct {
+	session    *session.Session
+	client     *s3.S3
+	uploader   *s3manager.Uploader
+	downloader *s3manager.Downloader
+	endpoint   string
+	callerTag  string
+}
+
+// S3 object store map with endpoint as the key to serve as cache
+var s3ConnectionMap = map[string]*s3ObjectStore{}
+
 // Create a bucket; don't return error if the bucket exists already
-func (s *s3Connection) createBucket(bucket string) error {
+func (s *s3ObjectStore) createBucket(bucket string) (err error) {
 	if bucket == "" {
 		return fmt.Errorf("empty bucket name for "+
 			"endpoint %s tag %s", s.endpoint, s.callerTag)
 	}
 
-	_, err := s.client.CreateBucket(
-		&s3.CreateBucketInput{Bucket: &bucket})
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("create bucket recovered for %s, with %v",
```

Comment:
Are you missing `return` in this deferred function? Or maybe you are relying on the named result argument.  I guess that's one of the reasons why I never liked to name my results arguments.  Very confusing especially when a function has many lines with many returns.  I would add at least after line 217 a `return`

Reply:
It is not missing a return statement.  The golang's deferred functions have the ability to change a named return value, and the above pattern appears to be a standard way.  See the JSON marshal function example at https://golang.org/src/encoding/json/encode.go#L322

Having said that, I can add a return line but that may be misleading because someone may interpret that the return is necessary.  While such a modified code compiles, the linter, however, complains that return value is not checked.

Reply:
Wow, that's interesting and very confusing.  Even the encode.go#L322 is small but very confusing. But I am fine with it if it works.

Reply:
Yes, this golang behavior is not intuitive but it serves a purpose.  From the spec: "For instance, if the deferred function is a function literal and the surrounding function has named result parameters that are in scope within the literal, the deferred function may access and modify the result parameters before they are returned. If the deferred function has any return values, they are discarded when the function completes. (See also the section on handling panics.)" https://golang.org/ref/spec#Defer_statements

Reply:
I added a comment to that defer function: "change the named return err value".